harness/determined/keras/callbacks.py

```diff
@@ -0,0 +1,410 @@
+            supports_tf_logs = getattr(cb, "supports_tf_logs", False)
+            cb.on_predict_end(polylogs(supports_tf_logs))
+
+    def _corrected_epoch_begin(self, epoch: int, logs: Optional[Dict] = None) -> None:
+        """The real on_epoch_begin call, which is guaranteed to happen at the right time."""
+        polylogs = _PolyLogs(logs)
+        for cb in self.callbacks:
+            supports_tf_logs = getattr(cb, "supports_tf_logs", False)
+            cb.on_epoch_begin(epoch, polylogs(supports_tf_logs))
+
+    def _corrected_epoch_end(self, epoch: int, logs: Optional[Dict] = None) -> None:
+        """The real on_epoch_end call, which is guaranteed to happen at the right time."""
+        polylogs = _PolyLogs(logs)
+        for cb in self.callbacks:
+            supports_tf_logs = getattr(cb, "supports_tf_logs", False)
+            cb.on_epoch_end(epoch, polylogs(supports_tf_logs))
+
+    def _check_validation_begin(self) -> None:
+        if self.state.want_validation_period_begin:
+            self.state.want_validation_period_begin = False
+            for cb in self.callbacks:
+                if isinstance(cb, Callback):
+                    cb.on_validation_period_begin(self.state.validation, {})
+
+    def _check_train_step_begin(self) -> None:
+        if self.want_train_step_begin:
+            self.want_train_step_begin = False
+            for cb in self.callbacks:
+                if isinstance(cb, Callback):
+                    cb.on_train_step_begin(self.state.total_batches, self.batches_requested, {})
+
+    def _check_epoch_begin(self) -> None:
+        if self.state.want_epoch_begin:
+            self.state.want_epoch_begin = False
+            self._corrected_epoch_begin(self.state.epoch, {})
+
+    # Must be triggered externally.
+    def set_batches_requested(self, batches_requested: int) -> None:
+        self.batches_requested = batches_requested
+
+    # Must be triggered externally.
+    def _validation_period_end(self, val_logs: Dict) -> None:
+        # Inject the latest training metrics automatically.
+        logs = {**val_logs, **self.state.latest_training_metrics}
+        logs.update(self.state.latest_training_metrics)
+        polylogs = _PolyLogs(logs)
+        for cb in self.callbacks:
+            if isinstance(cb, Callback):
+                supports_tf_logs = getattr(cb, "supports_tf_logs", False)
+                cb.on_validation_period_end(self.state.validation, polylogs(supports_tf_logs))
+        self.state.validation += 1
+        self.state.want_validation_period_begin = True
+
+    # Must be triggered externally.
+    def _train_step_end(self, metrics: Dict) -> None:
+        self.state.latest_training_metrics = metrics
+        for cb in self.callbacks:
+            if isinstance(cb, Callback):
+                cb.on_train_step_end(self.state.total_batches, metrics)
+        self.want_train_step_begin = True
+
+    # Must be triggered externally.
+    def _get_state(self) -> Dict:
```

Comment:
question: are you trying to save callbacks for all ranks here? Don't think it's necessary to introduce this complexity yet as we don't do this PyTorch and it has never been requested by anyone.

Reply:
ok, we can leave this out